### PR TITLE
Fix handling an empty list of versions

### PR DESCRIPTION
**Pull request by abhishekrb19**
This patch addresses this review [comment](https://github.com/apache/druid/pull/16141#discussion_r1531541234) by treating null and an empty list of versions differently. An empty collection passed should return no results, rather than ignoring it and returning all results. 

#### Changes
- Update the condition to handle the null and empty versions differently
- Clarify javadocs on the behavior of empty versions
- Add tests for the empty versions in different classes, which would fail without this patch

This PR has:

- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java [hunk 1]
```diff
@@ -927,7 +929,7 @@ private static void bindColumnValuesToQueryWithInCondition(
       final SQLStatement<?> query
   )
   {
-    if (CollectionUtils.isNullOrEmpty(values)) {
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java:930]: It is okay to return here on empty as well since if `values` is empty, we won't be doing any binding anyway.

**Review by kfaraz — APPROVED** [review]
Thanks for addressing this!

File: server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java (hunk 1 above)
Comment (abhishekrb19) [reply server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java:930]: True. I just kept it for consistency with its sibling method `getParameterizedInConditionForColumn()`.

**Comment by abhishekrb19** [comment]
Thanks for the quick review, @kfaraz!
